File: 프로그래머스/lv2/131704. 택배상자/택배상자.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

class Solution {
    public int solution(int[] order) {
        int answer = 0;
        int idx = 1;
        int stack = 1;
        Stack<Integer> st = new Stack<>();
        Queue<Integer> q = new LinkedList<>();
        for(int number : order){
            q.add(number);
        }
        st.add(1);
        while(!q.isEmpty()){
            if(q.peek() == idx){
                q.poll();
                st.pop();
                answer++;
                if(!st.isEmpty()) idx = st.peek();
                else idx = -1;
            }
            else if(idx > order.length) break;
            else if(idx > q.peek()) break;
            else{
                stack++;
                st.add(stack);
                idx = stack;
            }

        }
        
        return answer;
    }
}

File: 프로그래머스/2/131704. 택배상자/택배상자.py
from collections import deque

def solution(order):
    answer = 0
    idx = 0
    stack = []
    queue = deque(range(1, len(order) + 1))  

    while queue or stack:
        if queue and queue[0] == order[idx]:
            queue.popleft()
            answer += 1
            idx += 1
        elif stack and stack[-1] == order[idx]:
            stack.pop()
            answer += 1
            idx += 1
        elif queue:
            stack.append(queue.popleft())
        else:
            break

    return answer